graphql_compiler/schema_transformation/merge_schemas.py

```diff
@@ -238,13 +316,348 @@ def _process_generic_type_definition(generic_type, schema_id, existing_scalars,
+    field_name = field_reference.field_name
+
+    # Error if the type is nonexistent (includes if type is an enum or scalar)
+    # TODO: Consider improving error message if type is a union or enum or scalar
+    if type_name not in type_name_to_definition:
+        raise InvalidCrossSchemaEdgeError(
+            u'Type "{}" specified in the field reference "{}" is not found '
+            u'in the merged schema.'.format(type_name, field_reference)
+        )
+
+    # Error if the type is in a wrong or nonexistent schema
+    if type_name_to_schema_id[type_name] != schema_id:
+        raise InvalidCrossSchemaEdgeError(
+            u'Type "{}" specified in the field reference "{}" is expected to be in '
+            u'schema "{}", but is instead bound in schema "{}"'.format(
+                type_name, field_reference, schema_id, type_name_to_schema_id[type_name]
+            )
+        )
+
+    # Error if the type doesn't have the expected field
+    type_definition = type_name_to_definition[type_name]
+    type_fields = type_definition.fields
+    if not any(field.name.value == field_name for field in type_fields):
+        raise InvalidCrossSchemaEdgeError(
+            u'Field "{}" is not found under type "{}" in schema "{}", as expected by the '
+            u'field reference "{}".'.format(
+                field_name, type_name, schema_id, field_reference
+            )
+        )
+
+
+def _check_field_types_are_matching_scalars(type_name_to_definition, scalars, cross_schema_edge):
+    """Check that stitched fields in a cross schema edge are of the same scalar type.
+
+    It is also legal for fields to be of a NonNull wrapped scalar type.
+
+    Args:
+        type_name_to_definition: Dict[str, (Interface/Object)TypeDefinition], mapping
+                                 name of types to their definitions
+        scalars: Set[str], names of all scalars in the merged_schema so far
+        cross_schema_edge: CrossSchemaEdgeDescriptor namedtuple, the edge that we check the
+                           validity of
+
+    Raises:
+        - InvalidCrossSchemaEdgeError if the cross schema edge stitches together fields that are
+          not of a scalar type, or stitched together fields that are of different scalar types
+    """
+    field_type_names = []
+
+    for direction, field_reference in (
+        (OUTBOUND_EDGE_DIRECTION, cross_schema_edge.outbound_field_reference),
+        (INBOUND_EDGE_DIRECTION, cross_schema_edge.inbound_field_reference),
+    ):
+        type_name = field_reference.type_name
+        field_name = field_reference.field_name
+
+        fields = type_name_to_definition[type_name].fields  # List[FieldDefinition]
+        field_type = None
+        for field in fields:
+            if field.name.value == field_name:
+                field_type = field.type
+                while isinstance(field_type, ast_types.NonNullType):  # strip NonNull
+                    field_type = field_type.type
```

Comment:
For `NonNullType` in the language/ast.py file, it's `.type`. For `GraphQLNonNullType` in the type/definition.py file, it's `.of_type`. So I can't directly use that helper here.

I can make a new helper somewhere instead of inlining.